### PR TITLE
Fix nsfw author

### PR DESCRIPTION
**Pull request by ghost**
capture d ecran 2015-07-24 a 18 51 48

### CODE REVIEW

**Comment by jhass** [comment]
Two commits?

**Comment by AugierLe42e** [comment]
Myeah, mistake, sry.

**Comment by denschub** [comment]
I fear you [completely busted the build](https://travis-ci.org/diaspora/diaspora/builds/72502278).

**Comment by AugierLe42e** [comment]
It seems like. I correcting that today ;)

**Comment by robbyoconnor** [comment]
I see see the CI build failing...

**Comment by AugierLe42e** [comment]
I know @robbyoconnor , I was just too tired to fix them yesterday evening. Should be ok, now.

**Comment by robbyoconnor** [comment]
:+1: :joy_cat:

**Comment by AugierLe42e** [comment]
That's ok, now ;)

**Comment by svbergerem** [comment]
Photos aren't covered by the nsfw shield anymore.

**Comment by AugierLe42e** [comment]
Ouch, didn't catch that. Then I have a problem. Since photos are rendered above post infos, I cannot cover th pictures and the message but not the post infos. The only way I see to solve that is modify the layout.

**Comment by AugierLe42e** [comment]
Or else, I can only use CSS. Here is the result:

screen shot 2015-07-26 at 09 36 34

**Comment by svbergerem** [comment]
Why don't we do this exactly like in the desktop version? `display: none` (or `class="hidden"`) for text + image and when clicking "show post" we remove the nsfw shield and show the text and the image again.

**Comment by robbyoconnor** [comment]
:+1:

**Comment by AugierLe42e** [comment]
> Why don't we do this exactly like in the desktop version? `display: none` (or `class="hidden"`) for text + image and when clicking "show post" we remove the nsfw shield and show the text and the image again.

Nice idea. I'll give a try.

**Comment by AugierLe42e** [comment]
Here with using @svbergerem 's idea:

covered:

screen shot 2015-07-26 at 10 58 29

uncovered:

screen shot 2015-07-26 at 11 00 00

**Comment by AugierLe42e** [comment]
With the shield under the post info and comments hidden:

covered:

screen shot 2015-07-26 at 11 57 44

uncovered:

screen shot 2015-07-26 at 11 57 59

**Comment by AugierLe42e** [comment]
@svbergerem : It's ok to review ;)

**Comment by svbergerem** [comment]
Doesn't work for reshares. :(

Maybe do the same thing for `app/views/reshares/_reshare.mobile.haml`, add the nsfw shield there too and only display the nsfw shield in `_stream_element.mobile.haml` if the post is not a reshare?

Since there were now a few issues I think it would be great if you could add some cukes for original posts and reshares and make sure that several things (images, text, polls) are not visible.

**Comment by svbergerem** [comment]
Oh, here is another idea:
* Add a new class, say `.nsfw-hidden` to those things that should be hidden in nsfw posts. You can do that for both reshares and original posts
* Add some class, say `.shield-active` to the `.stream_element` and set `display: none` for all `.nsfw-hidden` elements inside of a `.shield-active`
* Clicking on “show post” removes the `.shield-active` class

Maybe also hide the location since we also hide it in the desktop view and could include content that is not appropriate for some users / places.

**Comment by AugierLe42e** [comment]
Done. I also changed `$(this).parents('.shield_wrapper').remove();` to `$(this).parents('.shield_wrapper').hide();`. This way we can add a feature to rehide the post like on desktop.

**Comment by AugierLe42e** [comment]
Reshare:

screen shot 2015-07-26 at 15 33 46

screen shot 2015-07-26 at 15 34 04

**Comment by AugierLe42e** [comment]
Why doesn't Travis start ? O__O

**Comment by jhass** [comment]
https://www.traviscistatus.com/

File: app/views/reshares/_reshare.mobile.haml [hunk 1]
```diff
@@ -4,17 +4,22 @@
 
 .reshare
   - if post
-    = render 'shared/photo_area', :post => post
+    .nsfw-hidden
+      = render "shared/photo_area", :post => post
```
Comment (svbergerem) [inline app/views/reshares/_reshare.mobile.haml]:
```ruby
= render "shared/photo_area", post: post
```
Please do that also for all other lines that you changed.

**Comment by AugierLe42e** [comment]
Done. Can you stock the previous build?

**Comment by goobertron** [comment]
I'm really pleased to see this being worked on at last. Thanks, Augier!

I prefer the version with the NSFW shield below the author details, as it makes sense not to show the reactions bar when the content is hidden.

By the way, that is a *truly* NSFW photo! ;-)

File: features/step_definitions/posts_steps.rb [hunk 2]
```diff
@@ -27,6 +27,22 @@
   user.post(:status_message, :text => text, :public => true, :to => user.aspect_ids)
 end
 
+Given /^"([^"]*)" has a public post with text "([^"]*)" and poll "([^"]*)"$/ do |email, text, question|
+  user = User.find_by_email(email)
```
Comment (jhass) [inline features/step_definitions/posts_steps.rb]: Use the newstyle finder, `User.find_by email: email`

File: features/step_definitions/posts_steps.rb [hunk 3]
```diff
@@ -27,6 +27,22 @@
...
+  poll = Poll.new(question: question)
+  poll.poll_answers.build(answer: "1")
+  poll.poll_answers.build(answer: "2")
+  post = user.post(:status_message, text: text, public: true, to: user.aspect_ids)
+  post.poll = poll
```
Comment (jhass) [inline features/step_definitions/posts_steps.rb]: Don't we have a factory (`spec/factories.rb`, `FactoryGirl`, factory_girl gem) for this?

File: features/step_definitions/stream_steps.rb [hunk 4]
```diff
@@ -14,6 +14,10 @@
   find(".toggle_nsfw_state", match: :first).click
 end
 
+When /^I toggle all nsfw posts$/ do
+  page.execute_script("$('a.toggle_nsfw_state').click()")
+end
```
Comment (jhass) [inline features/step_definitions/stream_steps.rb]: Do we really have to use a `execute_script` for this?

File: features/step_definitions/posts_steps.rb [hunk 5]
```diff
@@ -27,6 +27,22 @@
...
+end
+
+Given /^"([^"]*)" has a public post with text "([^"]*)" and location "([^"]*)"$/ do |email, text, loc|
+  user = User.find_by_email(email)
+  location = Location.new(coordinates: loc)
+  post = user.post(:status_message, text: text, public: true, to: user.aspect_ids)
+  post.location = location
```
Comment (jhass) [inline features/step_definitions/posts_steps.rb]: See above.

**Comment by AugierLe42e** [comment]
> I prefer the version with the NSFW shield below the author details, as it makes sense not to show the reactions bar when the content is hidden.

This was a request of @svbergerem . I don't really care. Both are ok for me.

> By the way, that is a truly NSFW photo! ;-)

Isn't it ? :p

**Comment by svbergerem** [comment]
The NSFW shield is below the details of the original author. In case of a reshare the "reshared by" line is below the NSFW shield because it is also below the content. The reactions bar is hidden for reshares and normal posts.

File: features/step_definitions/stream_steps.rb (hunk 4 above)
Comment (AugierLe42e) [inline features/step_definitions/stream_steps.rb]: Unless you can indicate me how to click multiple values with Capybara. I did a quick search. I didn't found. the `find` method doesn't let me click multiple links. I tried `find(".toggle_nsfw_state").each{ |elt| elt.click }` I got `Capybara::Ambiguous`.

Comment (svbergerem) [inline features/step_definitions/stream_steps.rb]: I think you are looking for [`all("...")`](http://www.rubydoc.info/github/jnicklas/capybara/master/Capybara/Node/Finders#all-instance_method)

**Comment by AugierLe42e** [comment]
Ok to merge for me.

File: features/mobile/not_safe_for_work.feature [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  And a user with email "laura@officeworkers.com" is connected with "tommy@pr0nking.com"
+  And "tommy@pr0nking.com" has a public post with text "Sexy Senators Gone Wild!" and a picture
+  And I sign in as "laura@officeworkers.com"
+  And I toggle all nsfw posts
+  And I follow "Reshare"
+  And I confirm the alert
+  And I go to the home page
+  Then I should not see "Sexy Senators Gone Wild!"
+  And I should not see any picture in my stream
+  And I should have 2 nsfw posts
```
Comment (jhass) [inline features/mobile/not_safe_for_work.feature]: `Scenario`s should be indented towards their `Feature`. Please extract the repetitive setup into a `Feature` level `Background`.

**Comment by AugierLe42e** [comment]
@jhass : Better like this ?

File: features/mobile/not_safe_for_work.feature [hunk 7]
```diff
@@ -0,0 +1,74 @@
+@javascript @mobile
+Feature: Not safe for work
+  Background:
+    Given following users exist:
+      | username    | email             |
+      | pr0n king   | tommy@pr0n.xxx    |
```
Comment (jhass) [inline features/mobile/not_safe_for_work.feature]: Am I missing something or do you never use this user?

Comment (AugierLe42e) [inline features/mobile/not_safe_for_work.feature]: [Am I not](https://github.com/diaspora/diaspora/pull/6245/files#diff-3bdf1f0e536df0353401430184957034R44) ?
Besides this, this whole file as been adapted from the desktop one. So I woudn't be able to justify every choice. I just know that it works, that's all.

Comment (svbergerem) [inline features/mobile/not_safe_for_work.feature]: `tommy@pr0nking.com` != `tommy@pr0n.xxx`

Comment (AugierLe42e) [inline features/mobile/not_safe_for_work.feature]: Right, and `laura@officeworkers.com` is not created either. How can this even work? No matter, I correct that, thank you.

Comment (denschub) [inline features/mobile/not_safe_for_work.feature]: When you are changing/writing any tests, please replace all domains with example.com. We've linked to some really stupid sites in the past and I don't wanna do that...

**Comment by AugierLe42e** [comment]
If there's no more remarks on this, then I'll do a rebase tonight and it can be merged.

File: features/mobile/not_safe_for_work.feature (hunk 7 above)
Comment (svbergerem) [inline features/mobile/not_safe_for_work.feature]: @AugierLe42e I agree with @denschub. Please change the domains to example.com while rebasing your PR.

Comment (jhass) [inline features/mobile/not_safe_for_work.feature]: note that labels beneath `example.com` and `example.org` are okay too, e.g. `office.example.com`.

Comment (AugierLe42e) [inline features/mobile/not_safe_for_work.feature]: Ok for me.

**Comment by AugierLe42e** [comment]
\o/

**Comment by svbergerem** [comment]
Thanks! :)

**Pull request merged**